### ISSUE TITLE
Expose Compilation.GetTypeByReflectionType and Assembly.GetTypeByReflectionType

### ISSUE THREAD

**Issue by fschmied** [dotnet/roslyn#4529]
While writing Roslyn-based code generators, we often ran into the situation of wanting to reference a well-known type from the generated code. For this, it would be nice if one could get an ``ITypeSymbol`` from a ``System.Type`` object. Roslyn already implements this functionality, but it's internal.

While ``Assembly.GetTypeByMetadataName`` can often, but not always, be used as a fallback, it has distinct disadvantages:
- ``GetTypeByMetadataName("System.Exception")`` is less type-safe than ``GetTypeByReflectionType(typeof(Exception))``;
- ``GetTypeByMetadataName(myType.FullName)`` often leads to wrong results (because Reflection formats some type names differently from the CLI metadata format), whereas ``GetTypeByReflectionType(myType)`` already implements the correct (and IMO non-trivial) formatting rules;
- ``GetTypeByMetadataName`` doesn't work for constructed generic types, array types, pointer types, etc, whereas ``GetTypeByReflectionType`` does. (Due to #3863, it's currently not even possible to work around this limitation.)

**Comment by tmat** [comment]
What do you expect the function to return if the given runtime type's declaring assembly (or an assembly of a type that is used to construct it) is not among the metadata references given to the compilation?

**Comment by fschmied** [comment]
@tmat I guess it should behave the same as ``GetTypeByMetadataName`` for reasons of symmetry.

(I believe that API returns ``null``. IMO, an exception would be a better contract, but in this case, symmetry would be more important to me.)

**Comment by tmat** [comment]
@fschmied The difference is that ```GetTypeByMetadataName``` does not specify assembly name while Type always has a declaring Assembly.

Note also that runtime unification policies apply when you load an assembly. In addition type forwarding happens. So it might not be straightforward to get what you want from the compilation.

**Comment by fschmied** [comment]
@tmat You're right, `GetTypeByMetadataName` doesn't specify an assembly name, however it still needs to deal with the possibility that the requested type isn't known by the `Compilation` and its referenced assemblies. Therefore, I think `GetTypeByReflectionType`'s behavior can be the same as `GetTypeByMetadataName`'s behavior in this case.

About the question of runtime policy and type forwarding: you're right, that makes the API more difficult to use than `GetTypeByMetadataName`. However, it can still be very valuable to have a quick way to go from a `Type` reference to an `ITypeSymbol` if you control these circumstances. For example, in my case, I know that the assembly in question doesn't have type forwarding and if someone adds a binding redirect, it's okay that the `Compilation` is required to reference the new version of the assembly as well. In summary, the existing implementation of `GetTypeByReflectionType` fits exactly what I need.
